### DISCUSSION TITLE
When logon as an admin user, WebView2 raised Element not found COMException during the initialization

### DISCUSSION THREAD

**Discussion by DreamlandForsure** [MicrosoftEdge/WebView2Feedback#3997] [Q&A]
Hi, WebView2 team,
Our application integrated the webview2 runtime for an internal embedded browser. Now end users reported an issue that the embedded browser is not working well. After debugging we found that the exception is actually raised at calling EnsureCoreWebView2Async() internally. We just follow the webview2 example to initialize the webview2 control.

_27/14:41:06 browsers   T100    :The exception raised as System.Runtime.InteropServices.COMException (0x80070490): Element not found. (Exception from HRESULT: 0x80070490)
27/14:41:06 browsers   T100    :   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
27/14:41:06 browsers   T100    :   at Microsoft.Web.WebView2.Core.CoreWebView2Environment.<CreateCoreWebView2ControllerAsync>d__45.MoveNext()
27/14:41:06 browsers   T100    :--- End of stack trace from previous location where exception was thrown ---
27/14:41:06 browsers   T100    :   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
27/14:41:06 browsers   T100    :   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
27/14:41:06 browsers   T100    :   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
27/14:41:06 browsers   T100    :   at Microsoft.Web.WebView2.Wpf.WebView2.<>c__DisplayClass27_0.<<EnsureCoreWebView2Async>g__Init|0>d.MoveNext()._

**The repro steps are as following -**
1. Our application is configured to run automatically at startup in Windows;
2. Log off(sign out) from the Windows 10 endpoint as an admin user;
3. Log on/back to the Windows 10 as an admin user;
Expected behavior: the application would pop up a window to display a initial web page. 
Actual behavior: the application pops up an error window which cannot display anything.

For the above scenario, there is no issue when use a non-admin user. And I also found there are below history tickets reported the similar issues. I confirmed that our application process is not the parent of the msedgewebview2.exe. My question is there any workaround to fix the issue? Or is there any fix developed for webview2 to fix the issue? Thanks.

https://github.com/MicrosoftEdge/WebView2Feedback/issues/1981
https://github.com/MicrosoftEdge/WebView2Feedback/issues/2780
_"We often found this error when running the application as admin and explorer is not in a good state… You could verify whether WebView thinks that the application is running elevated, and it has to make usage of explorer to launch its processes by observing in Task Manager whether the Edge WebView2 processes (msedgewebview2.exe) are shown as child processes to the application process. If the WebView2 processes are not shown as child processes of the application process, it indicates that WebView2 thinks that the application is running elevated."_

**Comment by victorthoang** [comment]
Hello @DreamlandForsure ,

Thanks for your inquiry! I've assigned this to a dev that might best answer your question.

**Comment by david-risney** [comment]
@LiangTheDev you've looked into this issue previously. Do you know if there is any solution or workaround?

**Comment by LiangTheDev** [comment]
The workaround for the normal case is to restart Windows Explorer using Task Manager.
In this case, it seems to be different. We are running at logon around the same time explorer is starting. So, I would suggest that we try to delay the webview creation a little, or retry webview creation after a delay, and don't display the UI before we know webview is created and working.

**Comment by DreamlandForsure** [comment]
Hi @victorthoang @david-risney , thanks for your prompt response and bridging the issue.
@LiangTheDev , your proposal sound reasonable. So is it possible or do you have a quick private of the fix which we can request a verification for our customer side?

**Comment by LiangTheDev** [comment]
Sorry for the confusion, I am expecting the app not WebView2 Runtime to make a change to work around the situation so that we have an improved end user experience.

**Comment by DreamlandForsure** [comment]
Hi @LiangTheDev , thanks for your kindly input. Let me discuss with our colleagues about if this is possible and how to implement the suggestion you proposed.

**Comment by byungmeo** [comment]
> The workaround for the normal case is to restart Windows Explorer using Task Manager. In this case, it seems to be different. We are running at logon around the same time explorer is starting. So, I would suggest that we try to delay the webview creation a little, or retry webview creation after a delay, and don't display the UI before we know webview is created and working.

Hi! @LiangTheDev 
The method of restarting Explorer.exe resolved issues that were not resolved for a particular user.

But I don't know how to solve the underlying problem. (Does the UI mean the UI of the entire app? Or is it limited to webview?)
And I wonder about the delay. In my case, I created the WebView2 object in Form_Load first and then ran await EnsureCoreWebView2Async(null).
How do I verify that WebView 2 is created and working?

```C#
private async void Form1_Load(object sender, EventArgs e) {
    webView_main = new WebView2() {
        Visible = true,
        Dock = DockStyle.Fill,
        Source = new Uri("https://www.naver.com")
    };
    await webView_main.EnsureCoreWebView2Async(null);
}
```

**Comment by sln162** [comment]
@byungmeo @LiangTheDev  In fact, this is the most troublesome problem for us. Several IPs report this error every day, and we did not expect so many at the beginning. The problem they mentioned should be because Windows has just logged in, so it needs to be delayed for a while to initialize, and most of the problems we encountered are concentrated in Win10 19044/19045, and the reason should be that the resource manager is not in a good state.
Our solution is to restart the explorer, for example: `taskkill /f /im explorer.exe&start explorer.exe` CMD command, and then wait for 10 seconds to initiate another reinitialization, if it still doesn't work, wait 20 seconds, 40 seconds, and so on. However, some users still cannot complete the initialization after 2 minutes.
We guess that the resource manager failed to restart, prompting the user to restart manually. Our program is started by the administrator, and WebView2 will be started through explorer.exe. Although I have suggested to the official many times, can a command be passed, similar to: `--allow-run-as-admin`, allowing administrator privileges to run WebView2, Instead of using the damn explorer.exe, alas, it is really painful to go on like this.

**Comment by LiangTheDev** [comment]
@byungmeo besides exception that would throw for await EnsureCoreWebView2Async, you could also use [WebView2.CoreWebView2InitializationCompleted Event](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.corewebview2initializationcompleted?view=webview2-dotnet-1.0.1462.37) to get notified about the WebView creation result.

**Comment by olegsavelos** [comment]
Is there any updates on this? I am facing similar problem where web view fails to start due to explorer state or in case its not running at all, agree @sln162 that there should be an option not to rely on explorer at will.

**Comment by byungmeo** [comment]
@olegsavelos 
It's not the best way, but I'm using a method that encourages users to install Edge Beta who occasionally contact me for issues with WebView 2 initialization or installation. I don't know why, but most of the issues have been resolved. I hope it helps!

**Comment by LiangTheDev** [comment]
We made some changes in Edge 112 to rely less on explorer state, that's probably why switching to beta improved. But we still depend on explorer being there for launching non elevated Edge WebView2 process.

**Comment by olegsavelos** [comment]
> We made some changes in Edge 112 to rely less on explorer state, that's probably why switching to beta improved. But we still depend on explorer being there for launching non elevated Edge WebView2 process.

In my case the process is fully elevated, the only problematic scenario is when explorer have not been started yet. 
Replicating this is quite easy, just kill explorer shell process and run application, com exception is raised (Element not found).

So how do I test beta ? Just install Edge beta version ?

**Comment by LiangTheDev** [comment]
I don't think running Edge beta would help as Edge 112 is already the current stable version. If explorer process is killed, I don't expect it to work. If you really want to try, you can install Edge beta and set environment variable WEBVIEW2_RELEASE_CHANNEL_PREFERENCE with value of 1 and then run the app.

**Comment by sln162** [comment]
@LiangTheDev  Last time I noticed your mention of the changes in version 112 on another issue, and from our backend data, we did see a significant reduction in the "element not found" error, but it has not completely disappeared. The issue still exists in version 112.0.1722.64, but at least we appreciate your attention and efforts. Will future versions of 112 or version 113 continue to optimize this issue?

Additionally, was this change made from the first version of 112? Or from which subversion did it start? Were there any changes related to this issue in the intermediate subversions? Because I want to limit the minimum runtime version and need a minimum version value, such as 112.0.1722.34.

**Comment by olegsavelos** [comment]
> I don't think running Edge beta would help as Edge 112 is already the current stable version. If explorer process is killed, I don't expect it to work. If you really want to try, you can install Edge beta and set environment variable WEBVIEW2_RELEASE_CHANNEL_PREFERENCE with value of 1 and then run the app.

Should we expect that at some point we can have web view running without explorer process? Even if its activated optionally.

**Comment by LiangTheDev** [comment]
All Edge 112 versions should have the same behavior. We still depend on explorer being there for WebView2 in elevated applications to work. There is still an issue we track, but I don't expect things to improve anytime soon.

**Comment by olegsavelos** [comment]
> All Edge 112 versions should have the same behavior. We still depend on explorer being there for WebView2 in elevated applications to work. There is still an issue we track, but I don't expect things to improve anytime soon.

Is there any way that we could determine that explorer is in an appropriate state in order for the web view to be launched ?
Basically correct me if I am wrong but it looks like right now this "Element not found" exception is thrown outside of our code so its impossible to probe and retry, perhaps there is a way we could try to initialize web view on application startup and make some retries in case of failure, right now I dont see any way to do this and apart from that we get unhandled exception in the application which is also undesirable.

**Comment by LiangTheDev** [comment]
I believe that the app can use try catch around EnsureCoreWebView2Async to handle the webview creation failure and then retry.

**Comment by olegsavelos** [comment]
> I believe that the app can use try catch around EnsureCoreWebView2Async to handle the webview creation failure and then retry.

I don't think this will work for me with the BlazorWebView I am using , perhaps I could probe for the com component in a loop with delay, any chance you could inform me what component I should try to create in order to see if explorer is fully initialized ?

**Comment by LiangTheDev** [comment]
@olegsavelos , the suggestion from BlazorWebView experts for how to retry:
Could you try completely removing/disposing the BlazorWebView and instantiating a completely new instance?
If you want a built-in API on BlazorWebView for re-initializing WebView2, they can submit a feature request in [dotnet/maui](https://github.com/dotnet/maui).

**Comment by sachinverma0005** [comment]
I'm encountering a comparable problem with my implementation of WebView2 in a WPF app, using version 1.0.1722.45 of WebView2 and version 1.3.173.55 of the runtime. Specifically, when the app is running in admin mode, I'm receiving an "Element not found" error, but it works fine in normal user mode.

Could you please provide some guidance on how to resolve this issue? Despite trying various approaches, I have been unable to find a solution.

**Comment by LiangTheDev** [comment]
@sachinverma0005 are you running in an environment where explorer is not running? It is currently a limitation of WebView2 and there is no workaround.

**Comment by sachinverma0005** [comment]
Yes, Explorer is not running.
But it's working fine in normal mode not in admin mode?

**Comment by LiangTheDev** [comment]
That's expected behavior. WebView2 processes can't run elevated and we depend on explorer being there for launching non elevated Edge WebView2 process. For apps running as standard user, we can simply use CreateProcess to create WebView2 processe, and therefore no dependency on explorer.

**Comment by olegsavelos** [comment]
> That's expected behavior. WebView2 processes can't run elevated and we depend on explorer being there for launching non elevated Edge WebView2 process. For apps running as standard user, we can simply use CreateProcess to create WebView2 processe, and therefore no dependency on explorer.

So I suppose the non-elevated launch is not a requirement but instead an security measure ? I really wish there was an option to allow web view process run elevated when security is not an issue.

**Comment by sln162** [comment]
#1981
#2772
These two open issues have been open for a long time, and have not received an official response. The cause of the error should be the same, which is related to the resource manager.

**Comment by LiangTheDev** [comment]
@sln162 #2772 is not the same as this issue. That was about using async APIs of host object injected into scripts. Closed #1981, as that is the same as this one.

**Comment by olegsavelos** [comment]
> That's expected behavior. WebView2 processes can't run elevated and we depend on explorer being there for launching non elevated Edge WebView2 process. For apps running as standard user, we can simply use CreateProcess to create WebView2 processe, and therefore no dependency on explorer.

Is there source code for this behavior available ? I would like to check the exact conditions when the browser process is launched under explorer. The idea is that perhaps it would be possible to create limited token when initializing web view and cause it not to require the explorer , perhaps this is something that development team should consider?

**Comment by LiangTheDev** [comment]
The code is basically what's in https://devblogs.microsoft.com/oldnewthing/20131118-00/?p=2643. We also try to use methods described in https://stackoverflow.com/questions/11169431.

**Comment by olegsavelos** [comment]
> The code is basically what's in https://devblogs.microsoft.com/oldnewthing/20131118-00/?p=2643. We also try to use methods described in https://stackoverflow.com/questions/11169431.

I see but what are conditions that determines if you should launch under explorer or not, I have tried impersonating the thread in multiple ways but still initialization of web view always fails with "Element not found" if explorer not running. Of course under the normal user that does not happen, so it would be nice to if you could clarify the conditions for launching under explorer as it would help me understand if impersonation route is possible or not.

**Comment by LiangTheDev** [comment]
You could try to create Edge/chrome browser process with --do-not-de-elevate --user-data-dir=<some-test-folder> to see whether that works. If that works, we should be able to also use the method to launch WebView2 processes.

**Comment by olegsavelos** [comment]
@LiangTheDev 

I tried it but no luck.
Here I have actually created an normal user and trying to impersonate him

```
  if (!AdvApi32.LogonUser("TestUser1", ".", "1234",
                    Win32API.Headers.Winbase.Defenitions.LOGON_TYPE.LOGON32_LOGON_INTERACTIVE,
                    Win32API.Headers.Winbase.Defenitions.LOGON_PROVIDER.LOGON32_PROVIDER_DEFAULT,
                    out IntPtr limitedToken))
                    throw new Win32Exception();

                await WindowsIdentity.RunImpersonatedAsync(new Microsoft.Win32.SafeHandles.SafeAccessTokenHandle(limitedToken), async () =>
                {
                    CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions("--do-not-de-elevate");
                    var env = await CoreWebView2Environment.CreateAsync(null, Environment.CurrentDirectory, options);
                    var controller = await env.CreateCoreWebView2ControllerAsync(handle);   
                });
```
Still if explorer not running "element not found", observing the user under which web view process runs reveals that its the current user and not the one being impersonated so I would assume you are not picking up the user from current thread ?

**Comment by LiangTheDev** [comment]
I see what you are asking. We are not picking up the impersonated token on the thread. WebView makes usage of multiple threads and lives longer than the impersonated period during webview2 initialization. And I am not sure whether parenting hwnds belongs to processes running as different users works or not.
You could change the code to launch `msedge.exe--do-not-de-elevate --user-data-dir=<some-test-folder>` to see whether the launched browser works. If the launched browser doesn't work, then it would not work. If the launched browser works, then it might be possible to make it work for WebView if we propagate the impersonation to all the places, though it would be very complicated.

**Comment by sachinverma0005** [comment]
I have created a seperate app, and opening that from my app (which is running in admin mode) with help of explorer.exe and that worked for me.
